packages/electron-updater/src/providerFactory.ts

```diff
@@ -1,45 +1,77 @@
 import { AllPublishOptions, BaseS3Options, BintrayOptions, GenericServerOptions, getS3LikeProviderBaseUrl, GithubOptions, newError, PublishConfiguration } from "builder-util-runtime"
+import { Provider } from "./main"
 import { AppUpdater } from "./AppUpdater"
 import { BintrayProvider } from "./BintrayProvider"
 import { GenericProvider } from "./GenericProvider"
 import { GitHubProvider } from "./GitHubProvider"
 import { PrivateGitHubProvider } from "./PrivateGitHubProvider"
 
-export function createClient(data: PublishConfiguration | AllPublishOptions, updater: AppUpdater) {
+export type ProviderFactoryFunction = (data: PublishConfiguration | AllPublishOptions, updater: AppUpdater) => Provider<any>
+
+interface ProviderFactoryList {
+  [key: string]: ProviderFactoryFunction
+}
+
+const providerList: ProviderFactoryList = {
+  github: createGitHubProvider,
+  spaces: createSpacesOrS3Provider,
+  s3: createSpacesOrS3Provider,
+  generic: createGenericProvider,
+  bintray: createBinTrayProvider
+}
+
+export function createClient(data: PublishConfiguration | AllPublishOptions, updater: AppUpdater): Provider<any> {
   // noinspection SuspiciousTypeOfGuard
   if (typeof data === "string") {
     throw newError("Please pass PublishConfiguration object", "ERR_UPDATER_INVALID_PROVIDER_CONFIGURATION")
   }
 
-  const httpExecutor = updater.httpExecutor
-  const provider = data.provider
-  switch (provider) {
-    case "github":
-      const githubOptions = data as GithubOptions
-      const token = (githubOptions.private ? process.env.GH_TOKEN || process.env.GITHUB_TOKEN : null) || githubOptions.token
-      if (token == null) {
-        return new GitHubProvider(githubOptions, updater, httpExecutor)
-      }
-      else {
-        return new PrivateGitHubProvider(githubOptions, token, httpExecutor)
-      }
-
-    case "s3":
-    case "spaces":
-      return new GenericProvider({
-        provider: "generic",
-        url: getS3LikeProviderBaseUrl(data),
-        channel: (data as BaseS3Options).channel || null
-      }, updater, provider === "spaces" /* https://github.com/minio/minio/issues/5285#issuecomment-350428955 */)
-
-    case "generic":
-      const options = data as GenericServerOptions
-      return new GenericProvider(options, updater, options.useMultipleRangeRequest !== false && !options.url.includes("s3.amazonaws.com"))
-
-    case "bintray":
-      return new BintrayProvider(data as BintrayOptions, httpExecutor)
-
-    default:
-      throw newError(`Unsupported provider: ${provider}`, "ERR_UPDATER_UNSUPPORTED_PROVIDER")
+  const providerFactoryFunction = providerList[data.provider]
+  if (!providerFactoryFunction) {
+    throw newError(`Unsupported provider: ${data.provider}`, "ERR_UPDATER_UNSUPPORTED_PROVIDER")
+  }
+  return providerFactoryFunction(data, updater)
+}
+
+export function registerProviderType(name: string, factoryMethod: ProviderFactoryFunction) {
```

Comment:
The extension point is here. See type `ProviderFactoryFunction` for the required function signature.